what4/src/What4/Protocol/VerilogWriter/AST.hs

```diff
@@ -366,13 +411,21 @@ addWire tp isSigned name e =
 
 -- | Create a fresh identifier by concatenating the given base with the
 -- current fresh identifier counter.
-freshIdentifier :: String -> VerilogM sym n Identifier
+freshIdentifier :: T.Text -> VerilogM sym n Identifier
 freshIdentifier basename = do
   st <- get
-  let x = vsFreshIdent st
-  put $ st { vsFreshIdent = x+1 }
-  return $ basename ++ "_" ++ show x
-
+  let used = vsUsedIdentifiers st
+  case basename `Set.member` used of
+    True ->
+      do let x = vsFreshIdent st
+             freshName = T.concat [basename, "_", fromString (show x)]
+         put $ st { vsFreshIdent = x+1
+                  , vsUsedIdentifiers = Set.insert freshName used
+                  }
+         return freshName
+    False ->
+      do put $ st { vsUsedIdentifiers = Set.insert basename used }
+         return basename
```

Comment:
Could reduce the duplication here:

```
st <- get
let used = vsUsedIdentifiers st
let nm = if basename `Set.member` used
         then T.concat [basename, "_", fromString $ show $ length used]
         else basename
put $ st { vsUsedIdentifiers = Set.insert nm used }
return nm
```
Which would also allow removal of `vsFreshIdent` (assuming this is it's primary use).  The downside is that the differentiation numbers wouldn't be monotonic, but they already skip across different basenames.

Reply:
Hmm. This is tempting. I'm a little worried that calculating the length of the used list over and over might be an issue in the case where we have hundreds of thousands of identifiers (which is quite realistic -- one of the main examples I'm testing on right now has around 30k, and it's far from the largest we might consider). That said, using a `Vector` might alleviate that problem (and reduce space usage along the way). I'll experiment with it.

Reply:
Oh, wait, it's a `Set` already, which has O(1) size calculation.